README.md
# Pixel SDK
Pixel SDK is a photo and video editing framework written in Swift.

<p align="center">
<img src="https://www.cdn.pixelsdk.com/assets/img/screenshots/sdk/library_1_border.jpg" alt="Screenshot" width="23.6%" height="auto" class="docs-screenshot"/>&nbsp;<img src="https://www.cdn.pixelsdk.com/assets/img/screenshots/sdk/edit_1_border.jpg" alt="Screenshot" width="23.6%" height="auto" class="docs-screenshot"/>&nbsp;<img src="https://www.cdn.pixelsdk.com/assets/img/screenshots/sdk/edit_7_border.jpg" alt="Screenshot" width="23.6%" height="auto" class="docs-screenshot"/>&nbsp;<img src="https://www.cdn.pixelsdk.com/assets/img/screenshots/sdk/edit_3_border.gifx" alt="Screenshot" width="23.6%" height="auto" class="docs-screenshot"/>
</p>

- [Features](#features)
- [Getting Started](#getting-started)
    * [CocoaPods](#cocoapods)
    * [Swift Package Manager](#swift-package-manager)
- [Restrict the SDK](#restrict-the-sdk)
    * [Images Only](#images-only)
    * [Videos Only](#videos-only)
    * [Square Content Only](#square-content-only)
    * [Portrait Content Only](#portrait-content-only)
    * [Library Only](#library-only)
    * [Camera Only](#camera-only)
- [Present the Editor](#present-the-editor)
- [Programmatic Editing](#programmatic-editing)
- [Export Media](#export-media)
    * [Image Exports](#image-exports)
    * [Video Exports](#video-exports)
    * [Encoding Settings](#encoding-settings)
- [Transcode Media](#transcode-media)
- [Write Custom Filters](#write-custom-filters)
- [Customize Colors](#customize-colors)
- [License](#license)

## Features
✅ Fully Customizable Filters with 40+ Included Filters and Visual Effects

✅ Vine Style Video Camera

✅ Auto-Saving Drafts

✅ Video Segment Composing, Trimming, Scaling, and Re-ordering

✅ Photo and Video Adjustments (Brightness, Vibrance, Saturation, Contrast, Exposure, Hue, Warmth, Sharpness, Gamma, Highlights, Shadows, Vignette)

✅ Cropping, Rotation, and Horizontal/Vertical Perspective Correction

✅ Adjust Individual Video Segments and Whole Video Composition

✅ Programmatic and Visual Editing

✅ Press and Hold Photo/Video to See Original While Editing

✅ Landscape, Portrait and or Square Content Support

✅ Swipe for Realtime Camera Filters

✅ Top Down Photo Mode

✅ Camera Brightness and Tap to Focus

✅ Share Directly to Facebook, Instagram or Twitter.

✅ Custom Colors and Dark Mode Support.

✅ Localized in 40+ [Languages](https://github.com/GottaYotta/PixelSDK/blob/master/LANGUAGES.md) and Dynamic Type Accessibility Support.

✅ Import Photos and Videos from DSLR Camera. Lightning to USB Cable Required.

✅ Direct GPU Access with Metal and [GPUImage3](https://github.com/BradLarson/GPUImage3).

✅ Transcode Video Files with Codecs, Filters, Trimming and more.

✅ RAW Images and 60fps 4K UHD Video Support. HEVC Support.

✅ Full [Documentation](https://www.pixelsdk.com/docs/latest/).

<br>

<p align="center">
<img src="https://www.cdn.pixelsdk.com/assets/img/screenshots/sdk/camera_1_border.jpg" alt="Screenshot" width="23.6%" height="auto" class="docs-screenshot"/>&nbsp;<img src="https://www.cdn.pixelsdk.com/assets/img/screenshots/sdk/edit_4_border.jpg" alt="Screenshot" width="23.6%" height="auto" class="docs-screenshot"/>&nbsp;<img src="https://www.cdn.pixelsdk.com/assets/img/screenshots/sdk/edit_5_border.jpg" alt="Screenshot" width="23.6%" height="auto" class="docs-screenshot"/>&nbsp;<img src="https://www.cdn.pixelsdk.com/assets/img/screenshots/sdk/edit_6_border.jpg" alt="Screenshot" width="23.6%" height="auto" class="docs-screenshot"/>
</p>

## Requirements
- iPhone or iPad
- iOS 11+
- Xcode 11.4+

## Getting Started

**Note:** Extensive sample code can be found in the [Xcode sample project](https://github.com/GottaYotta/PixelSDK/archive/master.zip).
<getting-started-supplement>

### CocoaPods

[CocoaPods](https://cocoapods.org/) is a dependency manager for iOS projects. To integrate PixelSDK into your Xcode project&nbsp;using&nbsp;CocoaPods, first verify you have at least Xcode 11.4 or greater installed.

Then, ensure you have the latest version of CocoaPods installed by running the following command:

```
$ sudo gem install cocoapods
```

Specify PixelSDK in your `Podfile`:

```
pod 'PixelSDK'
```

Run the following command from within your project directory:

```
$ pod install
```

### Swift Package Manager

[Swift Package Manager](https://developer.apple.com/documentation/swift_packages/adding_package_dependencies_to_your_app) is a dependency manager built into Xcode. To integrate PixelSDK into your Xcode project&nbsp;using&nbsp;Swift Package Manager, first verify you have the latest version of Xcode installed.

In the Xcode menu bar select File > Add Packages and enter the following repository URL into the search bar:

```
https://github.com/GottaYotta/PixelSDK.git
```

For Dependency Rule select `Branch` and `master`. Then press Add Package.

Note: We do not recommend setting the Dependency Rule to a major version because it will prevent you from receiving critical bug fixes in the future.

### Setup

Include the following lines in your application `Info.plist`:

```xml
<key>NSPhotoLibraryUsageDescription</key>
<string>Photo access is needed so you can select photos from your library.</string>
<key>NSMicrophoneUsageDescription</key>
<string>Microphone access is needed so you can record video with sound.</string>
<key>NSCameraUsageDescription</key>
<string>Camera access is needed so you can take photos.</string>
```
Present the SDK in response to a user action, for example, clicking a button. The default primary filters and adjustment filters will be used. The SDK will support both photo and video of any dimension with access to both the camera and library.

```swift
import PixelSDK
```
```swift
let container = ContainerController()
container.editControllerDelegate = self

let nav = UINavigationController(rootViewController: container)
nav.modalPresentationStyle = .fullScreen

self.present(nav, animated: true, completion: nil)
```

Also implement its delegate method. This delegate method will be called when the Next button in the EditController is pressed. In response you should either dismiss the UINavigationController or push a new controller on. The below example pushes a blank controller on. Then use the provided  `session` parameter to [export your photo or video](#export-media) at your own convenience.
```swift
extension ViewController: EditControllerDelegate {

    func editController(_ editController: EditController, didFinishEditing session: Session) {
        let controller = UIViewController()

        editController.navigationController?.pushViewController(controller, animated: true)
    }
}
```

[Generate an API key](https://www.pixelsdk.com/dashboard/api-keys/) and specify it in your  `application(_, didFinishLaunchingWithOptions:)` of your App Delegate. <span style="display: none;">The following [pricing options](https://www.pixelsdk.com/#pricing) are available for your API key.</span> Without an API key, image and video exports will include a watermark. Keep your API key private.

```swift
import PixelSDK

func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {

    PixelSDK.setup("YOUR API KEY")

    return true
}
```

Optionally, specify a maximum video duration. The default maximum video duration is 80 seconds.

 ```swift
 // Set the maximum video duration to 3 minutes.
 PixelSDK.shared.maxVideoDuration = 60*3
 ```
 
 That's it! You now have full access to the SDK.

## Restrict the SDK
### Images Only

The below example presents the SDK with only support for images. The user will have access to both the library and camera.
```swift
// Show only the library and photo camera modes in the tab bar
let container = ContainerController(modes: [.library, .photo])
container.editControllerDelegate = self

// Include only images from the users photo library
container.libraryController.fetchPredicate = NSPredicate(format: "mediaType == %d", PHAssetMediaType.image.rawValue)
// Include only images from the users drafts
container.libraryController.draftMediaTypes = [.image]

let nav = UINavigationController(rootViewController: container)
nav.modalPresentationStyle = .fullScreen

self.present(nav, animated: true, completion: nil)
```

### Videos Only

The below example presents the SDK with only support for videos. The user will have access to both the library and camera.
```swift
// Show only the library and video camera modes in the tab bar
let container = ContainerController(modes: [.library, .video])
container.editControllerDelegate = self

// Include only videos from the users photo library
container.libraryController.fetchPredicate = NSPredicate(format: "mediaType == %d", PHAssetMediaType.video.rawValue)
// Include only videos from the users drafts
container.libraryController.draftMediaTypes = [.video]

let nav = UINavigationController(rootViewController: container)
nav.modalPresentationStyle = .fullScreen

self.present(nav, animated: true, completion: nil)
```

### Square Content Only

The below example presents the SDK with only support for creating square photos and videos. The user will have access to both the library and camera.
```swift
let container = ContainerController()
container.editControllerDelegate = self

// Only allow square content from the library cropper
container.libraryController.previewCropController.aspectRatio = CGSize(width: 1, height: 1)
// Only allow square content from the camera controller
container.cameraController.aspectRatio = CGSize(width: 1, height: 1)

let nav = UINavigationController(rootViewController: container)
nav.modalPresentationStyle = .fullScreen
self.present(nav, animated: true, completion: nil)
```

### Portrait Content Only

The below example presents the SDK with only support for creating portrait photos and videos. The user will have access to both the library and camera.
```swift
let container = ContainerController()
container.editControllerDelegate = self

// Only allow portrait content from the library cropper
container.libraryController.previewCropController.aspectRatio = CGSize(width: 3, height: 4)
// Only allow portrait content from the camera controller
container.cameraController.aspectRatio = CGSize(width: 3, height: 4)

let nav = UINavigationController(rootViewController: container)
nav.modalPresentationStyle = .fullScreen
self.present(nav, animated: true, completion: nil)
```

### Library Only

The below example presents the SDK with only support for the library.
```swift
let container = ContainerController(mode: .library)
container.editControllerDelegate = self

let nav = UINavigationController(rootViewController: container)
nav.modalPresentationStyle = .fullScreen
self.present(nav, animated: true, completion: nil)
```

### Camera Only

The below example presents the SDK with only support for the video camera.
```swift
let container = ContainerController(mode: .video)
container.editControllerDelegate = self

let nav = UINavigationController(rootViewController: container)
nav.modalPresentationStyle = .fullScreen
self.present(nav, animated: true, completion: nil)
```

## Present the Editor

The camera and library can be bypassed by presenting the [EditController](https://www.pixelsdk.com/docs/latest/Classes/EditController.html) with your own media.

<p align="center">
<img src="https://www.cdn.pixelsdk.com/assets/img/screenshots/sdk/edit_1_border.jpg" alt="Screenshot" width="23.6%" height="auto" class="docs-screenshot"/>&nbsp;<img src="https://www.cdn.pixelsdk.com/assets/img/screenshots/sdk/edit_2_border.jpg" alt="Screenshot" width="23.6%" height="auto" class="docs-screenshot"/>&nbsp;<img src="https://www.cdn.pixelsdk.com/assets/img/screenshots/sdk/edit_4_border.jpg" alt="Screenshot" width="23.6%" height="auto" class="docs-screenshot"/>&nbsp;<img src="https://www.cdn.pixelsdk.com/assets/img/screenshots/sdk/edit_3_border.gifx" alt="Screenshot" width="23.6%" height="auto" class="docs-screenshot"/>
</p>

### Images
The below example presents the EditController with an image named "test_image" and sets the initial primary filter to [Wilshire](https://www.pixelsdk.com/docs/latest/Classes/SessionFilterWilshire.html).

```swift
let image = UIImage(named: "test_image")!

let session = Session(image: image)
session.image!.primaryFilter = SessionFilterWilshire()

let editController = EditController(session: session)
editController.delegate = self

let nav = UINavigationController(rootViewController: editController)
nav.modalPresentationStyle = .fullScreen
self.present(nav, animated: true, completion: nil)
```
### Videos
The below example presents the EditController with two AVAssets named "test.mov" and "test2.mp4" and sets the initial primary filter to [Sepulveda](https://www.pixelsdk.com/docs/latest/Classes/SessionFilterSepulveda.html). These two assets will become segments of the video in their respective order.

You can also manually pass a renderSize into the Session initializer. For more information see [Session](https://www.pixelsdk.com/docs/latest/Classes/Session.html) documentation.

```swift
let asset1 = AVAsset(url: Bundle.main.url(forResource: "test", withExtension: "mov")!)
let asset2 = AVAsset(url: Bundle.main.url(forResource: "test2", withExtension: "mp4")!)

let _ = Session(assets: [asset1, asset2], sessionReady: { (session, error) in
    guard let session = session,
        let video = session.video else {
        print("Unable to create session: \(error!)")
        return
    }
    
    // Set the initial primary filter to Sepulveda
    video.primaryFilter = SessionFilterSepulveda()

    let editController = EditController(session: session)
    editController.delegate = self

    let nav = UINavigationController(rootViewController: editController)
    nav.modalPresentationStyle = .fullScreen
    self.present(nav, animated: true, completion: nil)
})
```

## Programmatic Editing
Images and videos can also be edited programmatically instead of visually. For more information see [Session](https://www.pixelsdk.com/docs/latest/Classes/Session.html) documentation.

For example, setting the primaryFilter of an image to [Wilshire](https://www.pixelsdk.com/docs/latest/Classes/SessionFilterWilshire.html):
```swift
session.image!.primaryFilter = SessionFilterWilshire()
```
Applying a [Brightness](https://www.pixelsdk.com/docs/latest/Classes/SessionFilterBrightness.html) filter to an image:
```swift
let brightnessFilter = SessionFilterBrightness()
brightnessFilter.normalizedIntensity = 0.2
session.image!.filters = [brightnessFilter]
```
Applying a [Saturation](https://www.pixelsdk.com/docs/latest/Classes/SessionFilterSaturation.html) filter to a whole video:
```swift
let saturationFilter = SessionFilterSaturation()
saturationFilter.normalizedIntensity = 0.3
session.video!.filters = [saturationFilter]
```
Applying a [Contrast](https://www.pixelsdk.com/docs/latest/Classes/SessionFilterContrast.html) filter to the first segment of a video:
```swift
let segment = session.video!.videoSegments.first!
let contrastFilter = SessionFilterContrast()
contrastFilter.normalizedIntensity = 0.2
segment.filters = [contrastFilter]
```
Trimming a segment so it [starts](https://www.pixelsdk.com/docs/latest/Classes/SessionVideoSegment.html#/c:@M@PixelSDK@objc(cs)SessionVideoSegment(py)trimStartTime) at one second in, with a [duration](https://www.pixelsdk.com/docs/latest/Classes/SessionVideoSegment.html#/c:@M@PixelSDK@objc(cs)SessionVideoSegment(py)trimDuration) of two seconds:
```swift
let segment = session.video!.videoSegments.first!
segment.trimStartTime = CMTime(seconds: 1, preferredTimescale: segment.duration.timescale)
segment.trimDuration = CMTime(seconds: 2, preferredTimescale: segment.duration.timescale)
```
Rotating the first segment of a video with [preferredTransform](https://www.pixelsdk.com/docs/latest/Classes/SessionVideoSegment.html#/c:@M@PixelSDK@objc(cs)SessionVideoSegment(py)preferredTransform):
```swift
let segment = session.video!.videoSegments.first!
segment.preferredTransform = .rotated180Degrees(segment.naturalSize)

CHANGELOG.md
# Changelog

### 11.3.2 - 8/2/22

&bull; Fixed issue where some videos could not be selected from the library

### 11.3.1 - 7/24/22

&bull; General bug fixes and stability improvements

### 11.3.0 - 6/28/22

&bull; Added delegate methods to `CameraControllerDelegate` for camera button presses

### 11.2.0 - 4/6/22

&bull; Added support for custom camera button<br>
&bull; If content is watermarked, use a generic watermark when app is published on App Store

### 11.1.4 - 3/18/22

&bull; General bug fixes and stability improvements

### 11.1.3 - 3/17/22

&bull; Prevent Xcode from automatically changing framework version when an app is distributed

### 11.1.2 - 3/15/22

&bull; Updated example

### 11.1.1 - 2/10/22

&bull; Fixed issue where editing media of certain sizes would cause the UI to freeze (e.g. selfies from the native Camera app)<br>
&bull; General bug fixes and stability improvements

### 11.1.0 - 1/19/22

&bull; Added Swift Package Manager support<br>
&bull; Fixed crash on iOS 15.2 that occurs when the user opens their Library for the first time

### 11.0.5 - 11/16/21

&bull; Fixed bug which would cause HDR content to appear washed out on iPhone 12 and iPhone 13

### 11.0.4 - 9/30/21

&bull; Fixed crash caused by iOS 15 UIView infinite layout loop<br>
&bull; Fixed crash caused by pressing back on EditController while photo is still being captured<br>
&bull; General bug fixes and stability improvements

### 11.0.3 - 8/11/21

&bull; Added ability to provide empty audio encoding settings dictionary for no audio on export<br>
&bull; Improved stability when device disk space is full

### 11.0.2 - 12/23/20

&bull; Fixed variable scope issue with image, video, and segment of  `SessionFilterControllerAdditions`

### 11.0.1 - 11/21/20

&bull; Added CocoaPods minimum version requirement to Podspec

### 11.0.0 - 11/21/20

&bull; Added simulator support for M1 Macs (arm64)<br>
&bull; Migrated GPUImage dependency to xcframework distribution for ABI stability and reduced compile time<br>
&bull; Fixed issue that would occasionally prevent audio playback on `EditController` after coming from `CameraController`

### 10.2.0 - 11/19/20

&bull; Added support for system language override with `languageCode` variable<br>
&bull; Fixed photo camera incompatibility with iPhone 12, iPhone 12 Pro and iPhone 12 Pro Max<br>
&bull; Fixed bug which prevented dates from being properly encoded on Arabic locale devices<br>
&bull; Fixed filter thumbnails hide/show animation on `CameraController`

### 10.1.2 - 10/10/20

&bull; Fixed static sound playback bug on iOS 14<br>
&bull; Fixed Xcode 12 missing availableVideoPixelFormatTypes when compiling on Simulator

### 10.1.1 - 8/28/20

&bull; Added missing localizations for `SessionFilterSpeed`

### 10.1.0 - 8/11/20

&bull; Added playback speed multiplier support<br>
&bull; Improved multithreading stability<br>
&bull; Fixed `LibraryController` albums animation<br>
&bull; Various bug fixes and stability improvements

### 10.0.0 - 7/23/20

&bull; Migrated to Metal from OpenGL<br>
&bull; Various bug fixes and stability improvements

### 9.1.1 - 7/9/20

&bull; Fix bug which could cause a `Session` to not immediately appear in the `savedSessions` array<br>
&bull; Added missing localizations for "Portrait Content Only", "Preparing" and "The camera is not supported in split view mode. Please switch to fullscreen mode."<br>
&bull; Various bug fixes and stability improvements

### 9.1.0 - 7/7/20

&bull; Added support for disabling cropping on `PreviewCropController`<br>
&bull; Fix missing error from `Session` sessionReady closure

### 9.0.0 - 7/3/20

&bull; Added portrait content only example<br>
&bull; Fix bug which would infrequently cause non-square content to be produced when in square content only mode<br>
&bull; Deprecated `squareCameraActive` in favor of `aspectRatio` on `CameraController`<br>
&bull; Various bug fixes and stability improvements

### 8.0.0 - 6/24/20

&bull; Added Xcode 12 and Swift 5.3 support<br>
&bull; Migrated to xcframework file from universal binary<br>
&bull; Fix bug which prevented compilation on Xcode 11.4

### 7.0.2 - 6/23/20

&bull; Fix bug which prevented the framework from being compiled on Swift 5.3

### 7.0.1 - 6/23/20

&bull; Fix bug which would occasionally cause media to layout incorrectly in the `PreviewCropController`<br>
&bull; Various bug fixes and stability improvements

### 7.0.0 - 6/22/20

&bull; Added iPad support<br>
&bull; Added tab bar customization in `EditController`<br>
&bull; Added support for custom `EditController` UIViewControllers<br>
&bull; Added real-time `PreviewController` updates after making programmatic changes to a `Session`<br>
&bull; Fix infrequent crash from adjusting intensity slider/rotation on video sessions<br>
&bull; Changed variables for primary/adjustment filters from closures to arrays<br>
&bull; Various bug fixes and stability improvements

### 6.1.0 - 6/15/20

&bull; Added support for custom `SessionFilter` UIViewControllers<br>
&bull; Added support for filtered thumbnails in the `requestThumbnail()` function<br>
&bull; Fix bug which prevented custom colors from working properly on iOS 11 and 12<br>
&bull; Various bug fixes and stability improvements

### 6.0.0 - 6/10/20

&bull; Fix bug which prevented user defined SessionFilters from being properly encoded/decoded<br>
&bull; Fix bug where CameraController/ContainerController delegate was called prematurely before photo capture had finished<br>
&bull; Added Objective-C support to ContainerControllerDelegate<br>
&bull; Renamed ContainerControllerDelegate function from `func containerController(_ containerController: ContainerController, didChangeMode mode: ContainerMode)` to `func containerControllerDidChangeMode(_ containerController: ContainerController)`<br>
&bull; Various bug fixes and stability improvements

### 5.5.0 - 6/10/20
&bull; Fix bug which prevented the EditController cancel/revert button from working properly in some cases<br>

Example/Example/ViewController.swift
//
//  ViewController.swift
//  PixelSDKExample
//
//  Created by Josh Bernfeld on 8/1/19.
//  Copyright © 2021 GottaYotta, Inc. All rights reserved.
//

import UIKit
import PixelSDK
import PhotosUI

class ViewController: UITableViewController {
    
    @IBOutlet var activityIndicatorView1: UIActivityIndicatorView!
    @IBOutlet var activityIndicatorView2: UIActivityIndicatorView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
    }
    
    @IBAction func fullEditorWithStandardFilters() {
        PixelSDK.shared.primaryFilters = PixelSDK.defaultStandardFilters
        
        let container = ContainerController()
        container.editControllerDelegate = self
        
        let nav = UINavigationController(rootViewController: container)
        nav.modalPresentationStyle = .fullScreen
        self.present(nav, animated: true, completion: nil)
    }
    
    @IBAction func fullEditorWithVisualEffectFilters() {
        PixelSDK.shared.primaryFilters = PixelSDK.defaultVisualEffectFilters
        
        let container = ContainerController()
        container.editControllerDelegate = self
        
        let nav = UINavigationController(rootViewController: container)
        nav.modalPresentationStyle = .fullScreen
        self.present(nav, animated: true, completion: nil)
    }
    
    @IBAction func fullEditorWithCustomFilters() {
        PixelSDK.shared.primaryFilters = [
            SessionFilterExample1(),
            SessionFilterExample2(),
            SessionFilterWilshire(),
            SessionFilterMontana(),
            SessionFilterSanVicente(),
            SessionFilterMelrose(),
            SessionFilterSepulveda(),
            SessionFilterLaCienega(),
            SessionFilterAbbotKinney(),
            SessionFilterMulholland(),
            SessionFilterSunset()
        ]
        
        let container = ContainerController()
        container.editControllerDelegate = self
        
        let nav = UINavigationController(rootViewController: container)
        nav.modalPresentationStyle = .fullScreen
        self.present(nav, animated: true, completion: nil)
    }
    
    @IBAction func onlyLibraryController() {
        let container = ContainerController(mode: .library)
        container.editControllerDelegate = self
        
        let nav = UINavigationController(rootViewController: container)
        nav.modalPresentationStyle = .fullScreen
        self.present(nav, animated: true, completion: nil)
    }
    
    @IBAction func onlyVideoCameraController() {
        let container = ContainerController(mode: .video)
        container.editControllerDelegate = self
        
        let nav = UINavigationController(rootViewController: container)
        nav.modalPresentationStyle = .fullScreen
        self.present(nav, animated: true, completion: nil)
    }
    
    @IBAction func onlyPhotoCameraController() {
        let container = ContainerController(mode: .photo)
        container.editControllerDelegate = self
        
        let nav = UINavigationController(rootViewController: container)
        nav.modalPresentationStyle = .fullScreen
        self.present(nav, animated: true, completion: nil)
    }
    
    @IBAction func squareContentOnly() {
        let container = ContainerController()
        container.editControllerDelegate = self
        
        // Only allow square content from the library cropper
        container.libraryController.previewCropController.aspectRatio = CGSize(width: 1, height: 1)
        // Only allow square content from the camera controller
        container.cameraController.aspectRatio = CGSize(width: 1, height: 1)
        
        let nav = UINavigationController(rootViewController: container)
        nav.modalPresentationStyle = .fullScreen
        self.present(nav, animated: true, completion: nil)
    }
    
    @IBAction func portraitContentOnly() {
        let container = ContainerController()
        container.editControllerDelegate = self
        
        // Only allow portrait content from the library cropper
        container.libraryController.previewCropController.aspectRatio = CGSize(width: 3, height: 4)
        // Only allow portrait content from the camera controller
        container.cameraController.aspectRatio = CGSize(width: 3, height: 4)
        
        let nav = UINavigationController(rootViewController: container)
        nav.modalPresentationStyle = .fullScreen
        self.present(nav, animated: true, completion: nil)
    }
    
    @IBAction func editControllerWithCustomImageSession() {
        let image = UIImage(named: "test_image")!
        
        let session = Session(image: image)
        
        // Initialize the EditController
        let editController = EditController(session: session)
        editController.delegate = self
        
        // Present the EditController
        let nav = UINavigationController(rootViewController: editController)
        nav.modalPresentationStyle = .fullScreen
        self.present(nav, animated: true, completion: nil)
    }
    
    @IBAction func editControllerWithCustomVideoSession() {
        // NOTE: The session video will automatically inherit its renderSize from the actualSize of the first segment (asset) unless you pass a renderSize parameter to the session initializer.
        let asset1 = AVAsset(url: Bundle.main.url(forResource: "test_2", withExtension: "mp4")!)
        let asset2 = AVAsset(url: Bundle.main.url(forResource: "test_1", withExtension: "mov")!)
        let asset3 = AVAsset(url: Bundle.main.url(forResource: "test_3", withExtension: "mp4")!)
        
        let _ = Session(assets: [asset1, asset2, asset3], sessionReady: { (session, error) in
            self.activityIndicatorView1.stopAnimating()
            self.activityIndicatorView1.isHidden = true
            
            guard let session = session,
                let video = session.video else {
                print("Unable to create session: \(error!)")
                return
            }
            
            // Set the initial primary filter to Sepulveda
            video.primaryFilter = SessionFilterSepulveda()
            
            // Apply a Brightness filter to the first segment
            let brightnessFilter = SessionFilterBrightness()
            brightnessFilter.normalizedIntensity = 0.2
            video.videoSegments.first!.filters = [brightnessFilter]
            
            // Initialize the EditController
            let editController = EditController(session: session)
            editController.delegate = self
            
            // Present the EditController
            let nav = UINavigationController(rootViewController: editController)
            nav.modalPresentationStyle = .fullScreen
            self.present(nav, animated: true, completion: nil)
        })
        
        self.activityIndicatorView1.startAnimating()
        self.activityIndicatorView1.isHidden = false
    }

    @IBAction func imageContentOnly() {
        // Show only the library and photo camera modes in the tab bar
        let container = ContainerController(modes: [.library, .photo])
        container.editControllerDelegate = self
        
        // Include only images from the users photo library
        container.libraryController.fetchPredicate = NSPredicate(format: "mediaType == %d", PHAssetMediaType.image.rawValue)
        // Include only images from the users drafts
        container.libraryController.draftMediaTypes = [.image]
        
        let nav = UINavigationController(rootViewController: container)
        nav.modalPresentationStyle = .fullScreen
        
        self.present(nav, animated: true, completion: nil)
    }
    
    @IBAction func videoContentOnly() {
        // Show only the library and video camera modes in the tab bar
        let container = ContainerController(modes: [.library, .video])
        container.editControllerDelegate = self
        
        // Include only videos from the users photo library
        container.libraryController.fetchPredicate = NSPredicate(format: "mediaType == %d", PHAssetMediaType.video.rawValue)
        // Include only videos from the users drafts
        container.libraryController.draftMediaTypes = [.video]
        
        let nav = UINavigationController(rootViewController: container)
        nav.modalPresentationStyle = .fullScreen
        
        self.present(nav, animated: true, completion: nil)
    }
    
    @IBAction func customViewControllerInTabBar() {
        let myController = UIViewController()
        myController.view.backgroundColor = .purple
        
        let container = ContainerController(modes: [.library, .photo, .video, .custom(title: "Custom", controller: myController)])
        container.editControllerDelegate = self
        let nav = UINavigationController(rootViewController: container)
        nav.modalPresentationStyle = .fullScreen
        self.present(nav, animated: true, completion: nil)
    }
    
    @IBAction func transcodeVideoWithFilters() {
        // NOTE: The session video will automatically inherit its renderSize from the actualSize of the first segment (asset) unless you pass a renderSize parameter to the session initializer.
        let asset1 = AVAsset(url: Bundle.main.url(forResource: "test_1", withExtension: "mov")!)
        let asset2 = AVAsset(url: Bundle.main.url(forResource: "test_2", withExtension: "mp4")!)
        let asset3 = AVAsset(url: Bundle.main.url(forResource: "test_3", withExtension: "mp4")!)
        
        let _ = Session(assets: [asset1, asset2, asset3], sessionReady: { (session, error) in
            guard let session = session,
                let video = session.video else {
                print("Unable to create session: \(error!)")
                return
            }
            
            // Mark the session as transient so it does not persist on disk/appear in the users drafts
            session.isTransient = true
            
            // Set the video frame rate to 60 fps
            video.frameDuration = CMTime(value: 1, timescale: 60)
            
            // Apply a Saturation filter to the first segment
            let saturationFilter = SessionFilterSaturation()
            saturationFilter.normalizedIntensity = 0.2
            video.videoSegments[0].filters = [saturationFilter]
            
            // Apply a Pixellate filter to the second segment
            let pixellateFilter = SessionFilterPixellate()
            video.videoSegments[1].filters = [pixellateFilter]
            
            // Write to an MP4 file with H.264 video encoding and stereo AAC audio encoding
            
            var acl = AudioChannelLayout()
            memset(&acl, 0, MemoryLayout<AudioChannelLayout>.size)
            acl.mChannelLayoutTag = kAudioChannelLayoutTag_Stereo
            
            let audioEncodingSettings: [String: Any] = [
                AVFormatIDKey: kAudioFormatMPEG4AAC,
                AVNumberOfChannelsKey: 2,
                AVSampleRateKey: AVAudioSession.sharedInstance().sampleRate,
                AVChannelLayoutKey: NSData(bytes:&acl, length:MemoryLayout<AudioChannelLayout>.size),
                AVEncoderBitRateKey: 96000
            ]
            
            let videoEncodingSettings: [String: Any] = [
                AVVideoCompressionPropertiesKey: [
                    AVVideoProfileLevelKey: AVVideoProfileLevelH264HighAutoLevel,
                    AVVideoH264EntropyModeKey: AVVideoH264EntropyModeCABAC],
                AVVideoCodecKey: AVVideoCodecType.h264
            ]
            
            VideoExporter.shared.export(video: video,
                                        fileType: .mp4,
                                        videoEncodingSettings: videoEncodingSettings,
                                        audioEncodingSettings: audioEncodingSettings,
                                        progress: { (progress) in
                print("Transcode progress: \(progress)")
            }, completion: { (error) in
                self.activityIndicatorView2.stopAnimating()
                self.activityIndicatorView2.isHidden = true
                
                if let error = error {
                    print("Unable to transcode video: \(error)")
                    return
                }
                
                print("Finished video transcode at URL: \(video.exportedVideoURL)")
                
                let controller = UIAlertController(title: "Success", message: "Your video has been transcoded", preferredStyle: .alert)
                controller.addAction(.init(title: "Ok", style: .cancel, handler: nil))
                self.present(controller, animated: true, completion: nil)
            })
        })
        
        self.activityIndicatorView2.startAnimating()
        self.activityIndicatorView2.isHidden = false
    }
}

// MARK: - EditControllerDelegate

extension ViewController: EditControllerDelegate {
    
    func editController(_ editController: EditController, didLoadEditing session: Session) {
        // Called after the EditController's view did load.
    }
    
    func editController(_ editController: EditController, didFinishEditing session: Session) {
        // Called when the Next button in the EditController is pressed.
        // Use this time to either dismiss the UINavigationController, or push a new controller on.
        
        let controller = PostController()
        controller.session = session
        
        editController.navigationController?.pushViewController(controller, animated: true)
    }
    
    func editController(_ editController: EditController, didCancelEditing session: Session?) {
        // Called when the back button in the EditController is pressed.
    }
}
